R/reactives.R

```diff
@@ -118,18 +127,54 @@ ReactiveValues <- R6Class(
       )
       invisible()
     },
+
     mset = function(lst) {
       lapply(base::names(lst),
              function(name) {
                self$set(name, lst[[name]])
              })
     },
+
     names = function() {
       .graphDependsOn(.getReactiveEnvironment()$currentContext()$id,
                       sprintf('names(%s)', .label))
       .namesDeps$register()
       return(ls(.values, all.names=TRUE))
     },
+
+    # Get a metadata value. Does not trigger reactivity.
+    getMeta = function(key, metaKey) {
+      # Make sure to use named (not numeric) indexing into list.
+      metaKey <- as.character(metaKey)
+      .metadata[[key]][[metaKey]]
+    },
+
+    # Set a metadata value. Does not trigger reactivity.
+    setMeta = function(key, metaKey, value) {
+      # Make sure to use named (not numeric) indexing into list.
+      metaKey <- as.character(metaKey)
+
+      if (!exists(key, envir = .metadata, inherits = FALSE)) {
+        .metadata[[key]] <<- list()
+      }
+
+      .metadata[[key]][[metaKey]] <<- value
+    },
+
+    # Mark a value as invalid. If accessed while invalid, a shiny.silent.error
+    # will be thrown.
+    invalidate = function(key) {
+      setMeta(key, "invalid", TRUE)
+    },
+
+    unInvalidate = function(key) {
```

Comment:
Hm, that's a tough one. One thing that's a little different about invalidating values this way is that the behavior becomes dependent on the order that the reactives execute (imagine that you access `input$x` in one reactive and invalidate it in another, but those two reactives don't depend on each other). So we're already in a bit of a murky area.

Right now I don't see any advantage to doing it either way. For simplicity then, I'd say we leave it as is. But if you can see reasons to go one way or the other, I'm definitely open to that.

Two things to note:
* I think the names `invalidate` and `invalidateReactiveValue` are too similar to the *other* kind of invalidation that we do, so we should change the name.
* @bborgesr this code throws `shiny.silent.errors` but isn't yet in master, so you may need to make changes to this code later related to `reactiveStop()`.